source/assign-options.js

```diff
@@ -1,27 +1,36 @@
-const extend = require('extend');
+const url = require('url');
+const mergeWith = require('lodash.mergewith');
+const cloneDeep = require('lodash.clonedeep');
 const is = require('@sindresorhus/is');
 
 module.exports = (defaults, options = {}) => {
-	const returnOptions = extend(true, {}, defaults, options);
+	return mergeWith(
+		{opts: cloneDeep(defaults)},
+		{opts: options},
+		customizer
+	).opts;
+};
 
-	if (Reflect.has(options, 'headers')) {
-		for (const [key, value] of Object.entries(options.headers)) {
-			if (is.nullOrUndefined(value)) {
-				delete returnOptions.headers[key];
-			}
-		}
+function customizer(objValue, srcValue) {
+	if (is.array(srcValue) || is.array(objValue)) {
+		return cloneDeep(srcValue);
```

Comment:
This might not be necessary. Maybe we can just pick the right target based on whether is.array(objValue). I was running short on time, so I did this to get it working.

I think we could ditch `cloneDeep` as a dep if that works.